api/analyzedphenotypes.genus.api.inc

```diff
@@ -0,0 +1,194 @@
+<?php
+/**
+ * @file
+ *
+ * API: GENUS.
+ * Manage property and operation pertaining to data (Genus) in chado.organism table.
+ */
+
+
+/**
+ * DATA SOURCE - chado.organism.
+ * Fetch all genus in chado.organism.
+ * Result is sorted by genus alphabetically.
+ *
+ * @return
+ *   Array, all unique genus.
+ *   0 if no records found.
+ */
+function ap_get_genus() {
+  // All genus.
+  $result = chado_query(
+    "SELECT genus FROM {organism} GROUP BY genus ORDER BY genus ASC"
+  );
+
+  return ($result->rowCount()) ? $result->fetchCol() : 0;
+}
+
+
+///////
+
+
+/**
+ * CONTROL/SWITCHBOARD.
+ * Pass a request to a corresponding function.
+ * @TODO: leave out when api reorganize/update is complete and use functions below.
+ */
+function analyzedphenotypes_genusprop($property = null, $parameter = null) {
+  if (is_null($property)) {
+    // Default.
+    $property = 'default';
+  }
+
+  // Process request.
+  // Default to all projects.
+  $command = array(
+    'default' => array(
+      'func'  => 'ap_get_genus',
+      'param' => null,
+      'opts'  => null,
+    ),
+    'project_genus' => array(
+      'func'  => 'ap_get_projectgenus',
+      'param' => array('value' => (isset($parameter['project_id'])) ? $parameter['project_id'] : $parameter['project_name']),
+      'opts'  => array('is_id' => (isset($parameter['project_id'])) ? TRUE : FALSE),
+    ),
+    'assign_genus' => array(
+      'func'  => 'ap_set_projectgenus',
+      'param' => array('project_id' => $parameter['project_id'], 'genus' => $parameter['genus']),
+      'opts'  => null,
+    ),
+    'sysvar_genus_set' => array(
+      'func'  => 'ap_get_activegenus',
+      'param' => null,
+      'opts'  => null,
+    ),
+  );
+
+  // Function name, parameter (project name key or name), additional options.
+  return call_user_func($command[$property]['func'], $command[$property]['param'], $command[$property]['opts']);
+}
+
+
+///////
+
+
+/**
+ * FUNCTION:
+ * Assign a genus to an experiment/project in chado.projectprop table.
+ * Dependencies: ap_get_variablenames() - systemvars.api.
+ *
+ * @param $identifiers
+ *   An array with the following keys:
+ *   - project_id : project id number, to which the genus will be linked to.
+ *   - genus : genus.
+ */
+function ap_set_projectgenus($identifiers) {
+  // Fetch the configuration for the term Genus from terms system variable set.
+  $sysvar = ap_get_variablenames(
+    array('variablename' => 'genus'),
+    array('set' => 'terms')
+  );
+
+  $sysvar_genus = variable_get($sysvar);
+
+  // Before relating genus to a project, test for any previous correlations.
+  $is_rel = chado_query(
+    "SELECT projectprop_id FROM {projectprop} WHERE value = :genus AND project_id = :project_id AND type_id = :sysvar_genus",
+    array(
+      ':genus'   => $identifiers['genus'],
+      ':project_id' => $identifiers['project_id'],
+      ':sysvar_genus' => $sysvar_genus,
+    )
+  )
+  ->fetchCol();
+
```

Comment:
We should think about whether this is what we want to do... This doesn't allow someone to change an incorrect genus assignment.

Reply:
True, the front-end does not support changing it either, also, if we allow it to change it is likely someone might change correctly assigned genus. Should we let admin do the linking? in the meantime I can add another option that says replace/overwrite = TRUE or FALSE, then when we have a clear solution we have the api ready.

Reply:
I like this solution :-) Make the API support it but think further about the front-end 👍